Add 2022-12-01 agenda and minutes (#83)
solid/notifications-panel: meetings/2022-12-01.md

>ACTION: eP to review.

* eP: Reviewed. We can use notification sender/receiver. Client/server terminology not appropriate here.
* AC: If we change to those terms, we should also update the various diagrams.
* eP: I can follow up updating the diagrams.
* SC: So we change to 5 classes of products as in Pavlik's suggestion
* TT: +1 should be major reduction in confusion
* AC: +1
* SC: +1
* RG: May need to review Terminology
* SC: +1 after update to classes-of-products
* eP: +1

ACTION: eP to make suggested change to PR.

ACTION: eP to update diagram if necessary as separate PR

* RG: Separate PR for diagrams would be better

### Add relationship between `object` and `topic` resources
URL: https://github.com/solid/notifications/pull/127

* SC: There are different approaches as mentioned in https://github.com/solid/notifications/issues/62#issuecomment-1332663243 but keeping `topic` may be simplest at this time.
* SC: Closes: [:bike: shed: Use `object` (closely aligned with AS) instead of `topic` to receive notifications about a resource](https://github.com/solid/notifications/issues/62).

> * SC: Aaron and I talked about it yesterday. If we introduce a notion of request activity, to make a request, we should use `object` instead of `topic`. The other option with JSON-LD context, which protects `object`, it would be to much to exclude `object` from protected.

Reviewer: ```suggestion
* SC: Aaron and I talked about it yesterday. If we introduce a notion of request activity, to make a request, we should use `object` instead of `topic`. The other option with JSON-LD context, which protects `object`, it would be too much to exclude `object` from protected.
```

> * ...: This PR just links to the *topic* definition when we define *object*, it's just editorial change.

Reviewer: ```suggestion
* ...: This PR just links to the *topic* definition when we define *object*; it's just editorial change.
```

>SC: We need to follow up on this after resolving some issues in solid/notifications (linked in the review). This should be published alongside the next release of the Notifications Protocol.

* SC: https://github.com/solid/notifications/issues/74 for revisiting `notify:source`
* SC: https://github.com/solid/vocab/pull/62#pullrequestreview-995801549 :
  `rdfs:isDefinedBy <https://solidproject.org/TR/websocket-subscription-2021#websocketsubscription2021-source> ;`
* SC: Should we be changing that if other subscription types may be using `notify:source`?

> * eP: source and target should be in Protocol.
> * eP: target is used in webhook too. AFAIK LDN uses `inbox` instead of `target`.
> * AC: As I recall, we defined source at the websocket layer because we introduced it there, if it is reused across many subscription types we can define it at the protocol level. Making sure that change in the protocol shouldn't be a major change.

Reviewer: ```suggestion
* eP: `source` and `target` should be in Protocol.
* eP: `target` is used in webhook too. AFAIK LDN uses `inbox` instead of `target`.
* AC: As I recall, we defined `source` at the websocket layer because we introduced it there; if it is reused across many subscription types we can define it at the protocol level. Making sure that change in the protocol shouldn't be a major change.
```

> * eP: subscription types are either source or target based. source appears on the subscription response while target appears on subscription request.
> * SC: We should update request statements, at least one topic and 0-1 target, the response would be 0-1 source.
> * SC: do we need multiple targets?

Reviewer: ```suggestion
* eP: subscription types are either `source`- or `target`-based. `source` appears on the subscription response while `target` appears on subscription request.
* SC: We should update request statements, at least one `topic` and 0-1 `target`, the response would be 0-1 `source`.
* SC: do we need multiple targets?
```

* eP: I think we should start with 0 or 1
* AC: with multiple we get into various edge cases like partial failure. We should just have one at most.
* RG: all features are coming immediately now, if we are in a hurry we can leave it for next month.
* eP: +1 not to rush it in
* eP: Link [ Clarify subscription start and end times #111 ](https://github.com/solid/notifications/issues/111)

> * SC: okay with startTime as feature and renaming expiration to endTime.

Reviewer: ```suggestion
* SC: Okay with `startTime` as feature and renaming expiration to `endTime`.
```

* RG: There are two start times, one for from when notifications should start coming, and second for when events should trigger notifications.
* SC: I don't understand this distinction or if important
* RG: I was thinking about flash sales, one can subscribe up front but don't get notifications before midnight.
* eP: Why not notifications just coming at midnight?

> * RG: There is a start time from which you are ready to receive notifications and then i want the start time when to send notifications. connection to open before hand.

Reviewer: ```suggestion
* RG: There is a start time from which you are ready to receive notifications, and then I want the start time when to send notifications. Connection to open before-hand.
```

* SC: I think this is a little over-engineered.

> * RG: server will give me a websocket url. i should be able to access at some point eg, 11 but don't want notifications until midnight.

Reviewer: ```suggestion
* RG: server will give me a websocket url. I should be able to access at some point, e.g., 11pm, but don't want notifications until midnight.
```

* SC: I don't think RG was focusing on required or optional
* RG: I think this would be optional and we can defer it to next version 
* SC: I think we should just change expiry to end time, if server advertises end time, we should advertise start in the same way.
* RG: I would prefer that servers tells me when it will be sending notifications.
* eP: IMO we should add inline issue linking to #111 so we communicate it as a feature under an active discussion
* SC:
  ```
  #subscription request:
  type: one
  topic: at least one
  target: zero or one (for target based subscriptions)
  feature: zero or one per feature (#notification-features)
    endTime (renamed from expiration -- proposed change)
    startTime (proposed feature)

  #subscription response:
  type: one
  topic: at least one
  source: zero or one (for source based subscriptions)
  sender: zero or one (for target based subscriptions)
  feature: zero or one per feature (#notification-features)
    startTime (zero or one)
    endTime (zero or one: renamed from expiration -- proposed to require in the response)
  ```
* RG: would start time be a feature? I think all would be features in the request.

> * SC: If we just are leaving it as a future, what if we just push start time as a feature. I don't think any server is obligated to support specific features.

Reviewer: ```suggestion
* SC: If we just are leaving it as a feature, what if we just push start time as a feature. I don't think any server is obligated to support specific features.
```

* eP: Are you suggesting to have two different features, expiration and start time?
* SC: We should rename expiration to end time.
* eP: +1 to adjust it this way
* RG: I think end time should be required in the response.
* eP: subscription server should include endTime if it uses one. if feature is offered. it should be communicated if they want to leave it open (omit) but if determined they include endTime.
* RG: Will client assume infinite, if missing?
* SC: should be xsd valid if there is a value.
* eP: for subscription type, we'd want to be able to unsubscribe. next release should.

> * SC: LDN uses inbox, in LDN the term target is used for the thing that advertises it inbox. I don't think notification protocol should add inbox. 

Reviewer: ```suggestion
* SC: LDN uses `inbox`. In LDN, the term `target` is used for the thing that advertises its inbox. I don't think notification protocol should add inbox. 
```

* eP: We have different namespaces we do have `ldp:inbox` and we discuss defining `notify:target`. Do you consider also defining `notify:target`?
* SC: https://solid.github.io/notifications/linkeddatanotifications-subscription-2021

> * eP: What benefit do you use in defining custom property for LDN subscription type? it could specify that `notify:target` points to an LDN inbox.

Reviewer: ```suggestion
* eP: What benefit do you see in defining custom property for LDN subscription type? It could specify that `notify:target` points to an LDN inbox.
```

> * SC: I don't. `notify:target` is fine for subscription request in LDN sub type. we used inbox to reuse terminology from LDN. didn't have the hindsight about `target` for some types of subscription types.

Reviewer: ```suggestion
* SC: I don't. `notify:target` is fine for subscription request in LDN subtype. we used `inbox` to reuse terminology from LDN. Didn't have the hindsight about `target` for some types of subscription types.
```

* SC: I don't mind changing `inbox` to `target` in LDN subscription type.
* SC: what about `webid`?

> * eP: I find the property name pretty bad, we could use DID later on.

Reviewer: ```suggestion
* eP: I find the property name pretty bad; we could use DID later on.
```

* SC: LDN defines `webid` as agent who is authenticated to send and request.
* SC: The term could be `senderid` or just `sender`.
* eP: +1 to sender.
* SC: Or maybe `agent`?
* eP: Too general. doesn't indicate how it relates to subscription response. `sender` expresses it much better.
* RG: Agree.
* SC: What is Webhooks specification using?

> * eP: 0.1 has custom auth, draft PR for 0.2 currently reuses `webid` from LDN

Reviewer: ```suggestion
* eP: 0.1 has custom auth; draft PR for 0.2 currently reuses `webid` from LDN.
```

> * SC: I commented on https://github.com/solid/notifications/issues/15 where Kevin was tracking allowing accepting multiple types of notifications, saying that when client finds out which types are available in the subscription request they just need to specify one.

Reviewer: ```suggestion
* SC: I commented on https://github.com/solid/notifications/issues/15 where Kevin was tracking allowing accepting multiple types of notifications; saying that when client finds out which types are available in the subscription request, they just need to specify one.
```

* eP: makes sense, that issue must have been created before we defined how subscription server advertises available subscription type.

> * RG: Do we want a default type or want to always require specifying type.

Reviewer: ```suggestion
* RG: Do we want a default type or want to always require specifying type?
```

* SC: That'd be rejected because it doesn't conform to the subscription model.
* eP: Then type should be exactly one.
* eP: state could be useful in request if client reconnects and want to resubscribe, I recall some prior conversation.
* RG: Should the features be nested?
* SC: If you look at [section 2.2](https://solid.github.io/notifications/protocol#subscription). What we talk in [section 5](https://solid.github.io/notifications/protocol#notification-features) ???
* SC: We should leave features as separate section.

ACTION: SC to PR above.

### Interchangeable use of notification channel type and subscription type
URL: https://github.com/solid/notifications/issues/71

### Release of the spec
URL: https://github.com/solid/notifications-panel/blob/main/meetings/2022-11-10.md#release-of-the-spec

>ACTION: SC to review https://hackmd.io/3sLhQvmKTVKQKoXeXLSTpA

>The resource from which we obtain information about how we can subscribe to notification and what "types" of notifications are available.

> * SC: This is useful clarification - whether specific types are listed or not e.g., see https://github.com/solid/specification/issues/191 for a list of stuff. Not that those are actually advertised but they are observed on some level in a pod.

Reviewer: ```suggestion
* SC: This is useful clarification — whether specific types are listed or not, e.g., see https://github.com/solid/specification/issues/191 for a list of stuff. Not that those are actually advertised, but they are observed on some level in a pod.
```

> * SC: I suggest that the Solid Notifications Protocol (SNP) leaves this undefined to allow healthy variance. The key part here is that a resource server will decide whether it offers any notification service for a given resource. There is also no need to set this limitation on the SNP spec.

Reviewer: ```suggestion
* SC: I suggest that the Solid Notifications Protocol (SNP) leave this undefined to allow healthy variance. The key part here is that a resource server will decide whether it offers any notification service for a given resource. There is also no need to set this limitation on the SNP spec.
```

>+ The resource/container on which we make a POST request to request a for new subscription

> >+ The resource (if we decide that a resource is created?) which is created upon a request for a new subscription. The requester can also subsequently GET on this resource and obtain the subscription specific information (see below).

Reviewer: ```suggestion
>+ The resource (if we decide that a resource is created?) which is created upon a request for a new subscription. The requester can also subsequently GET this resource and obtain the subscription-specific information (see below).
```

* SC: Is this https://github.com/solid/notifications/issues/36 ? And touched on in https://github.com/solid/notifications/issues/111#issuecomment-1257700934 ?

> >In one of the meetings the possibility of having resource specific discovery resource was raised. Presumably, if the requester uses that resource specific information then subscription would then be specific to that resource. How do you support connections to multiple resources then? (Issue #22: note cxres issues)

Reviewer: ```suggestion
>In one of the meetings the possibility of having resource specific discovery resource was raised. Presumably, if the requester uses that resource-specific information, then subscription would then be specific to that resource. How do you support connections to multiple resources then? (Issue #22: note cxres issues)
```

* SC: Obviously can't. Request can be sent but I presume the subscription server would only recognise the resource that's valid or maybe just reject the request altogether.

...

> >I am against mandating Activity Streams as an exclusive format. I would rather have language that says "at a minimum" AS2 but provide flexibility to serve other formats (this is important to me for creating a unified abstraction that can span storages other than solid pods in a performant way).

Reviewer: ```suggestion
>I am against mandating Activity Streams as an exclusive format. I would rather have language that says "at a minimum" AS2 but provide flexibility to serve other formats. (This is important to me for creating a unified abstraction that can span storages other than solid pods in a performant way.)
```
